### ISSUE TITLE
Command does not recieve STDIN (musl?)

### ISSUE THREAD

**Issue by user18130814200115-2** [OmarCastro/cmd-polkit#9]
Hi, I am trying to run cmd-polit on alpine linux aarch64 which uses musl libc.

On my setup, while the program compiles and runs without issues, no command passed with the -c flag receives standard input, though the program does try to pass it:
```
Vrbos:auth_dialog_data_write_to_channel:writing to command stdin: {"action":"request password","prompt":"Password: ","message":"Authentication is needed to run `/bin/echo 1' as the super user","polkit action":{"id":"org.freedesktop.policykit.exec","description":"Run a program as another user","message":"Authentication is required to run a program as another user","vendor name":"The polkit project","vendor url":"http://www.freedesktop.org/wiki/Software/polkit/","icon name":""}}
```
After printing this message (with verbose enabled) the program freezes.

I looked into `g_spawn_async_with_pipes` which is responsible for sending this information, though I couldn't find anything about this call conflicting with musl libc or any other parts of my setup.

For further info: I run open-rc (no systemd), consolekit, seatd, and dbus (no elogind) and I tried zsh, bash, and ash while troubleshooting.

I also compiled and ran both the latest git and latest release.

Do you know what is going on here?

**Comment by user18130814200115-2** [comment]
Context: for testing i used just tofi as the program defined by -c, though i tried with various programs and wrappers. This is almost certainly not the problem

**Comment by OmarCastro** [comment]
Hello, since I am using arch linux, everything is built with glibc and could not compile it due to dependencies being built with glibc. I have one question can you run the unit tests with `meson test -C build ` and show the stdout of the command?
It may help me debug the problem while I find a suitable environment to test this problem

**Comment by user18130814200115-2** [comment]
Thanks for taking the time to help :)

I ran the tests as you asked. The first one fails `1/4 test-suite-unit                    ERROR           0.06s   killed by signal 5 SIGTRAP`

It seems to be `not ok / request messages / request message request password is escaped correctly - GLib-GObject-FATAL-CRITICAL: g_object_unref: assertion 'G_IS_OBJECT (object)' failed
Bail out!`

Also, just in case it is useful:

```
C compiler for the host machine: cc (gcc 14.2.0 "cc (Alpine 14.2.0) 14.2.0")
C linker for the host machine: cc ld.bfd 2.43.1
Host machine cpu family: aarch64
Host machine cpu: aarch64
Found pkg-config: YES (/usr/bin/pkg-config) 2.3.0
Run-time dependency glib-2.0 found: YES 2.80.4
Run-time dependency json-glib-1.0 found: YES 1.8.0
Run-time dependency polkit-agent-1 found: YES 125
Run-time dependency gtk+-3.0 found: YES 3.24.43
```

Full testlog:
```
Log of Meson test suite run on 2024-09-03T22:58:42.640997

Inherited environment: CHARSET=UTF-8 COLORFGBG='0;15' COLORTERM=truecolor DBUS_SESSION_BUS_ADDRESS=unix:path=/tmp/dbus-aq93xJADrl,guid=27bc8b2774bdfffa59c83e8f66d77162 DISPLAY=:0 EDITOR=vim HOME=/home/user KONSOLE_DBUS_SERVICE=:1.68 KONSOLE_DBUS_SESSION=/Sessions/1 KONSOLE_DBUS_WINDOW=/Windows/1 KONSOLE_VERSION=240800 LANG=C.UTF-8 LANGUAGE='' LC_COLLATE=C LOGNAME=user MUSL_LOCPATH=/usr/share/i18n/locales/musl PAGER=less PATH=/home/user/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin PROFILEHOME='' PWD=/home/user/Documents/cmd-polkit-0.3.0/Build QT_QPA_PLATFORM=wayland QT_QPA_PLATFORMTHEME=qt6ct SEATD_SOCK=/run/seatd.sock SHELL=/bin/ash SHELL_SESSION_ID=2c87810c4257422bb76f7a1e81680379 SHLVL=4 TERM=xterm-256color USER=user WAYLAND_DISPLAY=wayland-1 WINDOWID=2 XCURSOR_SIZE=24 XCURSOR_THEME=capitaine-cursors-dark XDG_CURRENT_DESKTOP=lxqt XDG_RUNTIME_DIR=/var/run/user/10000 XDG_SESSION_COOKIE=pine64-pinebookpro-1725395298.309559-1651859770 XDG_SESSION_TYPE=wayland OLDPWD=/home/user/Documents/cmd-polkit-0.3.0 NNN_TRASH=2 _=/usr/bin/meson 

==================================== 1/4 =====================================
test:         test-suite-unit
start time:   20:58:42
duration:     0.09s
result:       killed by signal 5 SIGTRAP
command:      ASAN_OPTIONS=halt_on_error=1:abort_on_error=1:print_summary=1 G_TEST_SRCDIR=/home/user/Documents/cmd-polkit-0.3.0 MESON_TEST_ITERATION=1 LC_ALL=C MSAN_OPTIONS=halt_on_error=1:abort_on_error=1:print_summary=1:print_stacktrace=1 MALLOC_PERTURB_=241 G_TEST_BUILDDIR=/home/user/Documents/cmd-polkit-0.3.0/Build UBSAN_OPTIONS=halt_on_error=1:abort_on_error=1:print_summary=1:print_stacktrace=1 /home/user/Documents/cmd-polkit-0.3.0/Build/cmd-polkit-agent-test-unit
----------------------------------- stdout -----------------------------------
TAP version 13
# random seed: R02S15a20b4ac5673a3ab315483173db1c32
1..25
# Start of  request messages  tests
ok 1 / request messages / all request messages are single line
# GLib-GIO-DEBUG: Using cross-namespace EXTERNAL authentication (this will deadlock if server is GDBus < 2.73.3)
not ok / request messages / request message request password is escaped correctly - GLib-GObject-FATAL-CRITICAL: g_object_unref: assertion 'G_IS_OBJECT (object)' failed
Bail out!
----------------------------------- stderr -----------------------------------

(test program exited with status code -5)
==============================================================================

==================================== 2/4 =====================================
test:         test-suite-it-parallel-mode
start time:   20:58:42
duration:     0.43s
result:       exit status 0
command:      ASAN_OPTIONS=halt_on_error=1:abort_on_error=1:print_summary=1 G_TEST_SRCDIR=/home/user/Documents/cmd-polkit-0.3.0 MESON_TEST_ITERATION=1 LC_ALL=C MSAN_OPTIONS=halt_on_error=1:abort_on_error=1:print_summary=1:print_stacktrace=1 G_TEST_BUILDDIR=/home/user/Documents/cmd-polkit-0.3.0/Build UBSAN_OPTIONS=halt_on_error=1:abort_on_error=1:print_summary=1:print_stacktrace=1 MALLOC_PERTURB_=22 /home/user/Documents/cmd-polkit-0.3.0/Build/cmd-polkit-agent-test-it-parallel-mode
----------------------------------- stdout -----------------------------------
TAP version 13
# random seed: R02S55c8d791e05cd9ca9351ab7e35280dd9
1..1
# Start of  polkit auth handler  tests
# GLib-GIO-DEBUG: Using cross-namespace EXTERNAL authentication (this will deadlock if server is GDBus < 2.73.3)
# GLib-GIO-DEBUG: Using cross-namespace EXTERNAL authentication (this will deadlock if server is GDBus < 2.73.3)
# GLib-GIO-DEBUG: Using cross-namespace EXTERNAL authentication (this will deadlock if server is GDBus < 2.73.3)
# DEBUG: received new line: {"action":"authenticate","password": "success"}
# 
# DEBUG: received new line: {"action":"authenticate","password": "success"}
# 
# DEBUG: received new line: {"action":"authenticate","password": "success"}
# 
# GLib-GIO-DEBUG: _g_io_module_get_default: Found default implementation local (GLocalVfs) for ‘gio-vfs’
ok 1 / polkit auth handler / CmdPkAgentPolkitListener initiate_authentication procedure serial mode
# End of  polkit auth handler  tests
----------------------------------- stderr -----------------------------------
parsing line
action authenticateparsing line
action authenticateparsing line
action authenticate
==============================================================================

==================================== 3/4 =====================================
test:         test-suite-it-serial-mode
start time:   20:58:42
duration:     1.01s
result:       exit status 0
command:      ASAN_OPTIONS=halt_on_error=1:abort_on_error=1:print_summary=1 G_TEST_SRCDIR=/home/user/Documents/cmd-polkit-0.3.0 MESON_TEST_ITERATION=1 LC_ALL=C MALLOC_PERTURB_=4 MSAN_OPTIONS=halt_on_error=1:abort_on_error=1:print_summary=1:print_stacktrace=1 G_TEST_BUILDDIR=/home/user/Documents/cmd-polkit-0.3.0/Build UBSAN_OPTIONS=halt_on_error=1:abort_on_error=1:print_summary=1:print_stacktrace=1 /home/user/Documents/cmd-polkit-0.3.0/Build/cmd-polkit-agent-test-it-serial-mode
----------------------------------- stdout -----------------------------------
TAP version 13
# random seed: R02S0670e837420c5488d357c3a3256a07bc
1..1
# Start of  polkit auth handler  tests
# GLib-GIO-DEBUG: Using cross-namespace EXTERNAL authentication (this will deadlock if server is GDBus < 2.73.3)
# GLib-GIO-DEBUG: Using cross-namespace EXTERNAL authentication (this will deadlock if server is GDBus < 2.73.3)
# GLib-GIO-DEBUG: Using cross-namespace EXTERNAL authentication (this will deadlock if server is GDBus < 2.73.3)
# DEBUG: received new line: {"action":"authenticate","password": "success"}
# 
# DEBUG: received new line: {"action":"authenticate","password": "success"}
# 
# DEBUG: received new line: {"action":"authenticate","password": "success"}
# 
# GLib-GIO-DEBUG: _g_io_module_get_default: Found default implementation local (GLocalVfs) for ‘gio-vfs’
ok 1 / polkit auth handler / CmdPkAgentPolkitListener initiate_authentication procedure serial mode
# slow test / polkit auth handler / CmdPkAgentPolkitListener initiate_authentication procedure serial mode executed in 0.97 secs
# End of  polkit auth handler  tests
----------------------------------- stderr -----------------------------------
parsing line
action authenticateparsing line
action authenticateparsing line
action authenticate
==============================================================================

==================================== 4/4 =====================================
test:         test-suite-it
start time:   20:58:42
duration:     1.84s
result:       exit status 0
command:      ASAN_OPTIONS=halt_on_error=1:abort_on_error=1:print_summary=1 G_TEST_SRCDIR=/home/user/Documents/cmd-polkit-0.3.0 MESON_TEST_ITERATION=1 LC_ALL=C MSAN_OPTIONS=halt_on_error=1:abort_on_error=1:print_summary=1:print_stacktrace=1 G_TEST_BUILDDIR=/home/user/Documents/cmd-polkit-0.3.0/Build UBSAN_OPTIONS=halt_on_error=1:abort_on_error=1:print_summary=1:print_stacktrace=1 MALLOC_PERTURB_=105 /home/user/Documents/cmd-polkit-0.3.0/Build/cmd-polkit-agent-test-it-command-exec
----------------------------------- stdout -----------------------------------
TAP version 13
# random seed: R02Sff31f325132bcf81a51cd5de073823ea
1..3
# Start of  polkit auth handler  tests
# GLib-GIO-DEBUG: Using cross-namespace EXTERNAL authentication (this will deadlock if server is GDBus < 2.73.3)
# DEBUG: received new line: {"action":"authenticate","password": "success"}
# 
ok 1 / polkit auth handler / CmdPkAgentPolkitListener initiate_authentication procedure success testing
# GLib-GIO-DEBUG: Using cross-namespace EXTERNAL authentication (this will deadlock if server is GDBus < 2.73.3)
# DEBUG: received new line: {"action":"cancel"}
# 
ok 2 / polkit auth handler / CmdPkAgentPolkitListener initiate_authentication procedure cancel testing
# GLib-GIO-DEBUG: Using cross-namespace EXTERNAL authentication (this will deadlock if server is GDBus < 2.73.3)
# DEBUG: received new line: {"action":"authenticate","password": "fail"}
# 
# DEBUG: received new line: {"action":"authenticate","password": "fail"}
# 
# DEBUG: received new line: {"action":"authenticate","password": "fail"}
# 
# DEBUG: received new line: {"action":"authenticate","password": "success"}
# 
ok 3 / polkit auth handler / CmdPkAgentPolkitListener initiate_authentication procedure fail retry testing
# slow test / polkit auth handler / CmdPkAgentPolkitListener initiate_authentication procedure fail retry testing executed in 1.68 secs
# End of  polkit auth handler  tests
----------------------------------- stderr -----------------------------------
parsing line
action authenticateparsing line
action cancelparsing line
action authenticateparsing line
action authenticateparsing line
action authenticateparsing line
action authenticate
==============================================================================


Summary of Failures:

1/4 test-suite-unit             ERROR           0.09s   killed by signal 5 SIGTRAP

Ok:                 3   
Expected Fail:      0   
Fail:               1   
Unexpected Pass:    0   
Skipped:            0   
Timeout:            0   

```

**Comment by user18130814200115-2** [comment]
I also tired building in a clean chroot, installing only the programs you use in CI, but no luck

**Comment by OmarCastro** [comment]
The integration test results shows that the commands are definitely receiving the STDIN and communicating to polkit through STDOUT. So the problem should not be there, did you try running one of the examples after installing them? you may need to install additional packages. For environments with X11 display manager, both examples work:

- rofi-example.sh: needs python and rofi installed
- zenity-jq-example.sh: needs zenity and jq installed

For environments with Wayland, only zenity-jq-example.sh work because rofi is X11 specific app (may work with xwayland but did not test it)

To test the examples do the following

- kill the polkit-agent process if there is one running already
- execute the example, assuming the required dependencies are installed
- run `pkexec  echo 1` in a terminal window

An authentication modal should appear, after putting the password correctly, it should print 1

**Comment by user18130814200115-2** [comment]
I feel very silly for imidiately thinking so compilcated, when the answer was quite simple: I was running the program wrong.

Instead of executing EG: `zentity-example.sh` I was running `cmd-polkit -s -c zenity-example.sh`. When this threw an error, I tried writing evermore minimal scripts to read STDIN, but always running them in the same way.

Thank you for your help, and sorry if I wasted your time :)
